File: Medium/145. Binary Tree Postorder Traversal/IterativeSolution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> result = new ArrayList<Integer>();

        TreeNode current = root;
        TreeNode previous = null;

        Stack<TreeNode> stack = new Stack<>();

        while (current != null || !stack.empty()) {
            if (current != null) {
                stack.push(current);
                current = current.left;
            } else {
                current = stack.peek();
                if (current.right == null || current.right == previous) {
                    result.add(current.val);
                    stack.pop();
                    previous = current;
                    current = null;
                } else
                    current = current.right;
            }
        }

        return result;
    }
}

File: 1. Easy/145. Binary Tree Postorder Traversal/iterative-solution.js
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var postorderTraversal = function (root) {
  let postorder = [],
      stack = [],
      current = root,
      previous = null;

  while (stack.length || current) {
    if (current) {
      stack.push(current);
      current = current.left;
    } else {
      current = stack[stack.length - 1];
      if (current.right && current.right != previous)
        current = current.right;
      else {
        postorder.push(current.val);
        stack.pop();
        previous = current;
        current = null;
      }
    }
  }

  return postorder;
};
